### PR TITLE
Udpate doc

### LINKED ISSUE #345: Revision of documentation
The documentation requires a good read and some updates.

- check for missing application and module documentation
- check for consistent interface description of modules
- add add hints regarding the installation of CORSIKA/sim_telarray (issue #172)

### PR DESCRIPTION
**Pull request by VictorBarbosaMartins**
Closes #345 

Hello Orel, Gernot,

finally (as promised) I am sending the PR with the documentation update. It has became much larger than I first thought and it is not yet complete. It was a lot of work for me, because on the fly I am learning about documentation and about the code structure itself.

I updated (chronologically):
- documentation of applications
- rst files
- classes in simtools (but not inside the subdirectories):
   - class methods
   - function parameters
   - return and raise statements
   - class parameters
   - class atributes

We are still missing:
- classes inside subdirectories in simtools

I ignored throughout the code the private functions (starting with an '_'), which I think is fine.

In order to be complete we will have to address those classes inside the subdirectories of simtools as well (plus your comments and suggestions to come). I will be during the morning in the workshop again and will answer several of your points later. Please feel free to work on this branch as well, I am opening the PR for that.

Thank you.
Best regards,
Victor

### CODE REVIEW

**Comment by GernotMaier** [comment]
Thanks for this - I will correct smallish typos directly without commenting.

File: docs/source/framework.rst [hunk 1]
```diff
@@ -12,48 +12,53 @@ Framework
 Configuration
```
Comment (GernotMaier) [inline docs/source/framework.rst:12]: This document needs a complete re-write given the new configurator. I will do this.

File: tests/integration_tests/test_applications.py [hunk 2]
```diff
@@ -66,7 +63,7 @@
             " --test",
         ]
     ],
-    "derive_mirror_rnda::psf_notuning": [
+    "derive_mirror_rnda::psf_measurement": [
```
Comment (GernotMaier) [inline tests/integration_tests/test_applications.py:66]: I am trying to understand the difference between `derive_mirror_rnda::psf_measurement` and `derive_mirror_rnda::psf_notuning`. Isn't this exactly the same?

File: tests/integration_tests/test_applications.py [hunk 3]
```diff
@@ -13,12 +13,20 @@
```
Comment (GernotMaier) [inline tests/integration_tests/test_applications.py:13]: What is the motivation to re-arrange the sequence of the tests? It makes it hard to compare! 

If there is no strong reason for this, could we go back to the original sequence? The possibility to track changes is important.

File: simtools/camera_efficiency.py [hunk 4]
```diff
@@ -541,7 +588,8 @@ def plot_cherenkov_efficiency(self):
 
         Returns
         -------
-        plt
+        fig
+            The figure
```
Comment (GernotMaier) [inline simtools/camera_efficiency.py:592]: I see the difficult of writing a good sentence here, but maybe we can discuss more what actually is returned? A matplotlob figure object? A pdf? An ascii figure? A path to a figure?

File: simtools/camera_efficiency.py [hunk 5]
```diff
@@ -574,7 +622,8 @@ def plot_nsb_efficiency(self):
...
```
Comment (GernotMaier) [inline simtools/camera_efficiency.py:626]: Same as previous comment.

File: simtools/camera_efficiency.py [hunk 6]
```diff
@@ -411,6 +434,11 @@ def calc_tel_efficiency(self):
     def calc_camera_efficiency(self):
         """
         Calculate the camera nominal efficiency including gaps (as defined in B-TEL-1170).
+
+        Returns
+        -------
+        cam_efficiency: float
+            Camera efficiency
```
Comment (GernotMaier) [inline simtools/camera_efficiency.py:441]: Is this a single value at one wavelength or an array for all wavelengths / offsets? Maybe we can be clearer here.

File: docs/source/applications.rst [hunk 7]
```diff
@@ -7,25 +7,28 @@ Applications
...
-These applications will be used as building blocks for the Simulation System Workflows.
+Applications are python scripts built on the :ref:`Library` that execute a simple, well defined
+task. These applications will be used as building blocks for the Simulation System Workflows.
 
 The application scripts can be found in gammasim-tools/applications.
 
-The applications expect a config file named config.yml in your gammasim-tools directory (unless specified differently
-using the function function config.setConfigFileName) . Please,
-follow the instructions in :ref:`Configuration` to setup your config file.
+The applications expect a config file named config.yml in your gammasim-tools directory (unless
```
Comment (GernotMaier) [inline docs/source/applications.rst:15]: This is a comment to myself that this paragraph needs an update given the new Configuration class

File: applications/print_array_elements.py [hunk 8]
```diff
@@ -29,12 +29,33 @@
 
     Example:
 
+    First get the array element list
```
Comment (GernotMaier) [inline applications/print_array_elements.py]: Here you didn't add `get ... from DB`, was that a conscious decision?

File: applications/derive_mirror_rnda.py [hunk 9]
```diff
@@ -174,23 +176,26 @@ def _parse(label):
         "--mirror_list",
         help=(
             "Mirror list file to replace the default one. It should be used if"
-            " measured mirror focal lengths need to be accounted"
+            " measured mirror focal lengths need to be accounted. It contains the following"
```
Comment (GernotMaier) [inline applications/derive_mirror_rnda.py:179]: I think we should not describe the columns and data fields of the input files and keep the help messages short (if possible to a single line).

**Comment by GernotMaier** [comment]
@VictorBarbosaMartins - I also suggest not to continue with this PR in the subdirectories for the modules but to do this in a separate PR. It will otherwise be hard to follow up!

File: applications/add_file_to_db.py [hunk 10]
```diff
@@ -6,15 +6,15 @@
     This application adds a file to the DB.
 
     The name and location of the file are required.
-    This application should complement the ones for updating parameters \
-    and adding entries to the DB.
+    This application should complement the ones for updating parameters, \
+    adding entries to the DB and getting file from the DB.
```
[the remark this answers is not in the document]
Comment (VictorBarbosaMartins) [reply applications/add_file_to_db.py]: Good point. I think we can indicate that the DB is chosen by the --db argument, then it will work for all DBs

File: applications/compare_cumulative_psf.py [hunk 11]
```diff
@@ -50,20 +50,32 @@
...
 
-    First, create an yml file named lst_pars.yml with the following content:
+    First, get the yml file named lst_pars_example.yml from the DB:
 
     .. code-block:: yaml
 
-        mirror_reflection_random_angle: '0.0075,0.15,0.035'
-        mirror_align_random_horizontal: '0.0040,28.,0.0,0.0'
-        mirror_align_random_vertical: '0.0040,28.,0.0,0.0'
+        python applications/get_file_from_db.py --file_name lst_pars_example.yml
```
[the remark this answers is not in the document]
Comment (VictorBarbosaMartins) [reply applications/compare_cumulative_psf.py:57]: Im not familiar with the application but as far as I understood, we wanted to test changing the parameters as well. In any case, as part of the example in the doc string we dont need to cover this case I think, I will remove it

[the remark this answers is not in the document]
Comment (VictorBarbosaMartins) [reply applications/compare_cumulative_psf.py:57]: Im not familiar with the application but as far as I understood, we wanted to test changing the parameters as well. In any case, as part of the example in the doc string we dont need to cover this case I think, I will remove it

File: simtools/camera_efficiency.py [hunk 12]
```diff
@@ -469,12 +508,20 @@ def calc_reflectivity(self):
...
     def calc_nsb_rate(self):
         """
         Calculate the NSB rate.
+
+        Returns
+        -------
+        nsb_rate
+            NSB rate
+        n1_sum
+            Sum of NSB counts
```
[the remark this answers is not in the document]
Comment (VictorBarbosaMartins) [reply simtools/camera_efficiency.py:524]: I agree adding the units makes more sense. I don't know why this function needs to return the counts, since the integral flux is calculated inside it. I havent found any use in the code for the returned counts, I will remove it

File: simtools/camera_efficiency.py [hunk 13]
```diff
@@ -469,12 +508,20 @@ def calc_reflectivity(self):
             [wl_now > 299 and wl_now < 551 for wl_now in self._results["wl"]]
         ]
         c2_sum = np.sum(c2_reduced_wl)
+        cher_spec_weighted_reflectivity = c2_sum / c1_sum / self._results["masts"][0]
```
[the remark this answers is not in the document]
Comment (VictorBarbosaMartins) [reply simtools/camera_efficiency.py:511]: I only assigned it in order to mention it in the docstring, otherwise it would look ugly (see below), hence, I suggest to keep the defined variable, or any other idea?
"""
        """
        Calculate the Cherenkov spectrum weighted reflectivity in the range 300-550 nm.

        Returns
        -------
        c2_sum / c1_sum / self._results["masts"][0]
            Cherenkov spectrum weighted reflectivity (300-550 nm)
        """
"""

File: simtools/camera_efficiency.py [hunk 14]
```diff
@@ -451,12 +484,18 @@ def calc_tot_efficiency(self, tel_effeciency):
         fill_factor = self._telescope_model.camera.get_camera_fill_factor()
 
         tel_effeciency_nsb = fill_factor * (n4_sum / (masts_factor * n1_sum))
+        tel_total_effeciency = tel_effeciency / np.sqrt(tel_effeciency_nsb)
```
[the remark this answers is not in the document]
Comment (VictorBarbosaMartins) [reply simtools/camera_efficiency.py]: same here, the assignment is only to facilitate the docstring, I wouldnt put any calculation in the docstring

File: docs/source/framework.rst (hunk 1 above)
Comment (VictorBarbosaMartins) [reply docs/source/framework.rst:12]: ok, thanks!

File: simtools/camera_efficiency.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (GernotMaier) [reply simtools/camera_efficiency.py:511]: Hmm, what does it add if you put a variable name into the docstring? What only setting:

```
Returns
--------
Float
   Cherenkov spectrum weighted reflectivity (300-550 nm)
```

The name of the parameter is not required, see https://numpydoc.readthedocs.io/en/latest/format.html#returns

File: tests/integration_tests/test_applications.py (hunk 2 above)
Comment (VictorBarbosaMartins) [reply tests/integration_tests/test_applications.py:66]: Indeed, a `--no_tuning`  argument was missing in the test

File: tests/integration_tests/test_applications.py (hunk 3 above)
Comment (VictorBarbosaMartins) [reply tests/integration_tests/test_applications.py:13]: I sorted it alphabetically, as it is displayed in my IDE. it makes more sense to me than chronologically. since we work on the tests throughout time, the chronology loses its meaning. There is the change to organize it according to the module, but the names of the applications do not always refer directly to a module, hence, it would be also difficult. My final suggestion is to keep it alphabetically.

File: simtools/camera_efficiency.py (hunk 4 above)
Comment (VictorBarbosaMartins) [reply simtools/camera_efficiency.py:592]: That is right. I now looked at the return statement of the respective function and it is a pyplot.fig, so I suggest I write there "Instance of pyplot.figure", "pyplot.figure" or "instance of Figure" (then pyplot is omitted) which one would be better?

File: simtools/camera_efficiency.py (hunk 6 above)
Comment (VictorBarbosaMartins) [reply simtools/camera_efficiency.py:441]: It is a single value (float) as indicated. It seems to me it is a wavelength-averaged mean. can you confirm that? it is hard to interpret when the variable is called "results"

File: applications/print_array_elements.py (hunk 8 above)
Comment (VictorBarbosaMartins) [reply applications/print_array_elements.py]: No, this was probably a lapse of memory. I will add it compatible to the previous case

File: tests/integration_tests/test_applications.py (hunk 3 above)
Comment (GernotMaier) [reply tests/integration_tests/test_applications.py:13]: You are right, the current sorting was kind of random. It just now almost impossible to compare the changes. I will have another look, but it makes reviewing much harder.

File: applications/derive_mirror_rnda.py (hunk 9 above)
Comment (VictorBarbosaMartins) [reply applications/derive_mirror_rnda.py:179]: I agree. the problem here is that there is no doctring for the function and if we just remove this explanation we lose it for good. I will create a docstring for the function and include the detailed information there and keep it small in the help message

File: simtools/camera_efficiency.py (hunk 6 above)
Comment (GernotMaier) [reply simtools/camera_efficiency.py:441]: Yes, this is correct.

File: applications/add_file_to_db.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (VictorBarbosaMartins) [reply applications/add_file_to_db.py]: Ok, I decided instead of it to add the article 'a' before it. it indicates that there are more DBs. Please indicate me again if you don't like it

File: applications/print_array_elements.py (hunk 8 above)
Comment (VictorBarbosaMartins) [reply applications/print_array_elements.py]: There is actually a difference between getting and adding a file to DB. In getting a file we don't need to indicate which db it is, we just search through all the available dbs, hence I will add here "from the DBs". On the other hand, in adding a file to DB  we have to indicate which DB, hence I used there "to a DB"

File: applications/derive_mirror_rnda.py (hunk 9 above)
Comment (VictorBarbosaMartins) [reply applications/derive_mirror_rnda.py:179]: Ok, so I reduced the size and did not add a docstring (I actually think it would not be necessary)

File: simtools/camera_efficiency.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (VictorBarbosaMartins) [reply simtools/camera_efficiency.py:511]: Ok, then I will pass only the name, but the consistency will be lost since in all the other functions the variable is returned

File: applications/add_file_to_db.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (GernotMaier) [reply applications/add_file_to_db.py]: Good

**Comment by VictorBarbosaMartins** [comment]
Ok, I pushed again, please have a second look. I will let you close the conversations in case you are fine with my changes.

File: simtools/camera_efficiency.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (GernotMaier) [reply simtools/camera_efficiency.py:486]: But you agree that this is clearer and simpler? We should rather apply in future this to the other methods.

File: simtools/camera_efficiency.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (GernotMaier) [reply simtools/camera_efficiency.py]: I don't think 'code should serve docstrings', it should be the other way around...

File: simtools/camera_efficiency.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (VictorBarbosaMartins) [reply simtools/camera_efficiency.py:486]: Indeed, it is not necessary to know the name of the returned variable since one has to define anyways a new one to save the returned value

File: simtools/camera_efficiency.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (VictorBarbosaMartins) [reply simtools/camera_efficiency.py]: oh, I forgot to change here too

File: simtools/camera_efficiency.py [hunk 15]
```diff
@@ -524,7 +522,7 @@ def calc_nsb_rate(self):
             * self._telescope_model.reference_data["nsb_reference_value"]["Value"]
             / nsb_integral
         )
-        return nsb_rate, n1_sum
+        return nsb_rate
```
Comment (GernotMaier) [inline simtools/camera_efficiency.py:524]: And you are sure that n1_sum is not sued when this function is called?

**Review by GernotMaier — APPROVED** [review]
Have a look at the one question, otherwise good!

File: simtools/camera_efficiency.py (hunk 15 above)
Comment (GernotMaier) [reply simtools/camera_efficiency.py:524]: But removing it just because we don't know the units is maybe not correct. I think @orelgueta knows the answer here.

Comment (VictorBarbosaMartins) [reply simtools/camera_efficiency.py:524]: I removed it because the function was returning a variable that was never use in the code.

Comment (orelgueta) [reply simtools/camera_efficiency.py:524]: It's a remnant from the first version of the code which I had locally (there it was used). Either Raul or I copied it but removed its usage. It's safe to remove.

File: tests/integration_tests/test_applications.py (hunk 3 above)
Comment (orelgueta) [reply tests/integration_tests/test_applications.py:13]: Wait, the order was not random!
I ordered it such that the first 5-10 applications were roughly the quickest ones. Originally we had the optics first which were the slowest tests first and I always had to wait 5-10 minutes just to pass the first ones. The tests were also bunched in Camera/Optics/DB/etc. That I actually liked a lot. Could we go back to the order before this change?

File: simtools/camera_efficiency.py (hunk 4 above)
Comment (orelgueta) [reply simtools/camera_efficiency.py]: `Instance of pyplot.figure` is best

**Review by orelgueta — COMMENTED** [review]
I couldn't find a way to remove Gernot's approval, so I comment here to please look first at my comment on the order of the applications before you merge.

File: tests/integration_tests/test_applications.py (hunk 3 above)
Comment (VictorBarbosaMartins) [reply tests/integration_tests/test_applications.py:13]: Ok, I just returned to the previous order

**Comment by VictorBarbosaMartins** [comment]
> I couldn't find a way to remove Gernot's approval, so I comment here to please look first at my comment on the order of the applications before you merge.

Maybe we should include that information in the developer guidelines regarding the order of the application tests

**Comment by orelgueta** [comment]
Thanks! Feel free to merge then.
